swift/apple/FirezoneNetworkExtension/Adapter.swift

```diff
@@ -139,85 +149,95 @@ class Adapter {
+    // Create the session
+    let session: Session
+    do {
+      session = try Session.newApple(
+        apiUrl: apiURL,
+        token: token.description,
+        deviceId: deviceId,
+        accountSlug: accountSlug,
+        deviceName: deviceName,
+        osVersion: osVersion,
+        logDir: logDir,
+        logFilter: logFilter,
+        deviceInfo: deviceInfoStr,
+        isInternetResourceActive: internetResourceEnabled
+      )
+    } catch {
+      throw AdapterError.connlibConnectError(String(describing: error))
+    }
 
-  /// Start the tunnel.
-  func start() throws {
-    Log.log("Adapter.start")
+    // Create channels - following Rust pattern with separate sender/receiver
+    let (commandSender, commandReceiver): (Sender<SessionCommand>, Receiver<SessionCommand>) =
+      Channel.create()
+    self.commandSender = commandSender
+
+    let (eventSender, eventReceiver): (Sender<Event>, Receiver<Event>) = Channel.create()
+    self.eventReceiver = eventReceiver
+
+    // Start event loop - owns session, receives commands, sends events
+    Task { [weak self] in
+      defer {
+        // ALWAYS cleanup, even if event loop crashes
+        self?.commandSender = nil
+        self?.eventReceiver = nil
+        Log.log("Adapter: Event loop finished, session dropped")
+      }
 
-    guard session == nil else {
-      throw AdapterError.invalidSession(session)
+      await runSessionEventLoop(
+        session: session,
+        commandReceiver: commandReceiver,
+        eventSender: eventSender
+      )
     }
 
-    callbackHandler.delegate = self
+    // Start event consumer - consumes events from receiver (Rust pattern: receiver outside)
+    Task { [weak self] in
+      guard let eventReceiver = self?.eventReceiver else { return }
 
-    Log.log("Adapter.start: Starting connlib")
-    do {
-      let jsonEncoder = JSONEncoder()
-      jsonEncoder.keyEncodingStrategy = .convertToSnakeCase
-
-      // Grab a session pointer
-      session = try WrappedSession.connect(
-        apiURL,
-        "\(token)",
-        "\(id)",
-        accountSlug,
-        DeviceMetadata.getDeviceName(),
-        DeviceMetadata.getOSVersion(),
-        connlibLogFolderPath,
-        logFilter,
-        internetResourceEnabled,
-        callbackHandler,
-        String(data: jsonEncoder.encode(DeviceMetadata.deviceInfo()), encoding: .utf8)!
-      )
-    } catch let error {
-      // `toString` needed to deep copy the string and avoid a possible dangling pointer
-      let msg = (error as? RustString)?.toString() ?? "Unknown error"
-      throw AdapterError.connlibConnectError(msg)
+      for await event in eventReceiver.stream {
+        await self?.handleEvent(event)
+      }
+
+      Log.log("Adapter: Event consumer finished")
     }
+
+    // Configure DNS and path monitoring
+    startNetworkPathMonitoring()
+
+    Log.log("Adapter.start: Session started successfully")
+  }
+
+  // Could happen abruptly if the process is killed.
+  deinit {
```

Comment:
I am not entirely sure how Swift's ownership model works but we should make sure to cancel all tasks and free all resources here.